### PR TITLE
fix(spec_decode): backport independent draft RNG from upstream #54282

### PR DESCRIPTION
**Pull request by MadeBy561**
**Ready for review/testing. Validation on the exact refreshed PR head is pending.**

## Why this matters

Speculative decoding lets a fast draft suggest tokens, then the main model
checks them. When rejecting a suggestion, the old code reused the random
numbers that helped choose it—like rerolling with loaded dice. That biased
the final token choices away from the main model's intended probabilities.

This backports **Giancarlo Delfin's upstream [vLLM #54282](https://github.com/vllm-project/vllm/pull/54282)**.
The original author is preserved. Our work is the JJ adaptation and additional
tests, not inventing the fix.

## How we proved it

We fed the actual GPU draft/rejection code a known three-token distribution.
For 131,072 trials at depth 3 / temperature 1:

| First-token frequency | A | B | C |
|---|---:|---:|---:|
| Intended | 10.00% | 60.00% | 30.00% |
| Original R25 | 10.16% | 63.04% | 26.80% |
| Backport | 10.13% | 59.95% | 29.92% |

The original error was far beyond expected sampling noise. The corrected
results fit the expected distribution across the tested cases.
**107 GPU tests passed**, including BF16 proposals and context-boundary cases.

Value: correct token probabilities—not a promise of higher acceptance or
better answers on every prompt.

[Raw distributions, runnable reproducer, full configuration and serving results](https://gist.github.com/MadeBy561/90b728c7b6b329895719a1e812d36534).

## Scope and testing

Prior rig: **GLM-5.3-Flash-NVFP4, TP4/DCP1, Marlin MTP3; 4× RTX PRO 6000
Blackwell Max-Q, 300 W/card, +6000 memory OC, zero core OC.**

Combined router + sampling results: Estonia 30/30, LAVD 30/30, GSM8K 64/64;
MMLU-Pro 49/64, tools 8/16. Normal thinking was retained. These do not isolate
this PR. Warm verifier speed was approximately unchanged; warm C1 output
and Sieve median were 4.3% and 3.9% lower, respectively.

Current-head pre-commit and diff checks passed. **Pending:** exact-head GPU
suite, isolated sampling-only model A/B, cache repeats and broader backend
testing. Production was untouched.

<details>
<summary>Technical reproduction, base and attribution</summary>

Base: `dev/jovian-judgement` at `f564dffe9`; commit `44e6766e3`.
Upstream: `fe755c88995ad468882517b6c4bdd60138d46a3a`.

Separates proposal/recovery Gumbel streams and sampling positions from
fixed/clamped forward positions. Preserves compact MTP prefill, MRoPE,
per-token logits caching and local DFlash2/DSpark callers.

```bash
.venv/bin/python -m pytest -q -p no:cacheprovider --confcutdir=tests/v1 \
  tests/v1/spec_decode/test_rejection_sampler_utils.py \
  tests/v1/worker/test_gpu_gumbel_sample.py \
  tests/v1/worker/test_gpu_autoregressive_speculator.py \
  tests/v1/worker/test_gpu_rejection_sampler_chunking.py \
  tests/v1/worker/test_gpu_rejection_sampler_i64.py
```

The attached probe needs `--explicit-sampling-positions` on the backport.
Duplicate checks found no equivalent JJ backport; the upstream fix is
already merged. Companions: #652 and [B12X #317](https://github.com/local-inference-lab/b12x/pull/317).

AI assistance: OpenAI Codex under MadeBy561's direction. Human line-by-line
review remains pending.

</details>

### CODE REVIEW

**Comment by voipmonitor** [comment]
## R27 integration validation

The change represented by this PR is included in the qualified, source-locked
GLM-5.3-Flash runtime
`voipmonitor/vllm:jovian-judgement-community-20260906-r27`
(`sha256:a298fe1cd207eaf97bd2ff2686716ed25b7009c09b36650eba732a4a7dc51512`).
The exact vLLM composition is mirrored at
[`voipmonitor/vllm:integration/glm53-r27-release-20260906`](https://github.com/voipmonitor/vllm/tree/integration/glm53-r27-release-20260906),
commit `63a82f8d323e8538cbe6f88ae1812a1c01577a0f`.

Qualification used four stock-clock RTX PRO 6000 Blackwell Workstation Edition
GPUs, TP4, a 4,096-token scheduler budget, 16 NCCL channels, a 2 MiB NCCL
buffer, and full plus piecewise CUDA graphs:

| Mode | DCP | 32K prefill | C1 output / steps | C8 output / steps |
|---|---:|---:|---:|---:|
| No speculation | 1 | 14,870 tok/s | 170.6 tok/s | 733.8 tok/s |
| MTP3 | 1 | 14,468 tok/s | 276.0 / 109.1 tok/s | 901.0 / 371.3 tok/s |
| MTP3 full CKV | 4 | 12,864 tok/s | 247.0 / 97.1 tok/s | 876.4 / 346.4 tok/s |
| DFlash2 K7 full CKV, NVFP4 KV | 4 | 12,633 tok/s | 198.0 / 81.2 tok/s | 645.5 / 260.8 tok/s |

FP8 no-speculation and NVFP4 DFlash2 external-cache configurations also passed
cold compute, vLLM prefix reuse, engine-driven RAM-L1 restore, full-process
filesystem-L2 restore, and block-checksum validation on all four ranks. An exact
81,576-token leading-instruction test reused 81,567 tokens when only the user
continuation changed.

This is an integration and regression gate, not an isolated attribution of the
aggregate throughput to this PR. The complete open-PR merge order and evidence
are recorded in #651.

**Comment by voipmonitor** [comment]
Included in `dev/jovian-judgement` through [this PR's individual merge](https://github.com/local-inference-lab/vllm/commit/bb779219b065ab7ac59c7f2f3f6a388655501990). The reviewed head and its contributor commits remain ancestors; the PR is merged and closed.

Source validation: replaying all 32 R35 review heads on the pinned base exactly reproduces the released Docker's vLLM tree; all 6,870 installed tracked files match. JJ additionally preserves Luke's DS4.1 work and #734. The final composition passed 247 focused checkpoint/scheduler, sampler/warmup and native GPU tests. This is combined-source evidence, not a fresh performance or full-model qualification for this individual PR.

Publication-history clarification: the individual merge linked above is in JJ's first-parent history. It replaces the receipt's archived wrapper-merge reference; GitHub's historical merge SHA may still identify that archive. See #731 for component review order and qualification limits.
